### ISSUE TITLE
deprecate bitwise `or` in favor of `bitor` (+ other operators that operate on bit representation)

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#8487]
this is super error prone, another gotcha to learn, can complicate generic code in some cases, and makes it harder to search.

These should be deprecated in favor of `bitor`, etc.
These are some examples:
```nim
proc `or`*(x, y: int): int {.magic: "BitorI", noSideEffect.}
proc `xor`*(x, y: int): int {.magic: "BitxorI", noSideEffect.}
proc `not`*[T: SomeUnsignedInt](x: T): T {.magic: "BitnotI", noSideEffect.}
proc `and`*[T: SomeUnsignedInt](x, y: T): T {.magic: "BitandI", noSideEffect.}
proc `xor`*[T: SomeUnsignedInt](x, y: T): T {.magic: "BitxorI", noSideEffect.}
proc `not`*(x: int): int {.magic: "BitnotI", noSideEffect.}
```

## proposal:
* every proc `foo` defined with `.magic: "BitXXX"` is deprecated and aliased to a proc named `bitfoo`
* eventually will become an error

**Comment by Bulat-Ziganshin** [comment]
Writing low-level bit-manipulation code will be even harder after that change. I would like to have C-style `| & ~` operators with proper priorities.

**Comment by timotheecour** [comment]
> Writing low-level bit-manipulation code will be even harder after that change

I disagree, it'll make it more obvious, and less easily confused for logical operators.

> I would like to have C-style | & ^ ~ operators with proper priorities.

I'd rather use more explicit names like `bitor` (which are also easier to search). I don't like reserving all these 4 symbols just for narrow (in scope) use case of bit operations, when simply using `bitor` + friends can take care of it. 

Here's a much better solution that doesn't pollute namespace:

```nim
withBitOperators: # defines C-style | & ^ ~
  let a = 1 | 2
```

**Comment by Bulat-Ziganshin** [comment]
It's more likely that I would need ptrMath+withBitOperators for en entire low-level module, but otherwise it's OK for me - C was developed 50 years ago and even at this time it explicitly targeted writing low-level code, so we don't need to copy its specific operator set in the modern application language :)

**Comment by Araq** [comment]
Why are these "super error prone" and a problem for "generic programming"? I don't like them too much either but have yet to see a bug that isn't caught by Nim's type checker on these.

**Comment by Araq** [comment]
@Bulat-Ziganshin We can add ``|`` etc to the ``bitops`` module. 

There is nothing "convenient" about ``x &= ~(1 << n)`` though, that's a ``clearBit(x, n)`` operation. C's specific operator set is just badly designed and it already was bad 50 years ago. ``setBit``, ``clearBit``, ``flipBit`` are more useful for low level programming and have to be remembered as specific patterns in C.

**Comment by Araq** [comment]
I'm teaching you that you have done some much low level programming that you don't see misdesigns anymore.

1. Ridiculous, you can map ``clearBit`` to the most efficient instruction sequence.
2. By that logic we don't need ``+`` because you can always use the more general ``-`` operator (a + b is the same as  a - (-b))

**Comment by Bulat-Ziganshin** [comment]
@Araq 
> We can add | etc to the bitops module.

I don't know where it should belong, but I would like to see `| & ^ ~ >> <<` plus ROL/ROR operations (f.e. as `<<<  >>>`) plus their op= counterparts with C-inspired priorities, compiled directly to corresponding C/C++ equivalents, so we can expect the same performance.

Even more interesting would be to have ability to define them myself. AFAIK, now the only barrier to do it is the lack of user-defined op priorities (sorry for reminding that too often). And there are other operators too that I would like to define in my programs.


>Why are these "super error prone" and a problem for "generic programming"?

BTW, one problem with existing and/or is their too-low priority. In my feeling, for bitOr/bitAnd it should be higher than +/-.


>I'm teaching you that you have done some much low level programming that you don't see misdesigns anymore.

I like to learn, but single-bit operation clearly can't extract multi-bit field from a number. Just a small sample (all masks are computed based on hash size and similar run-time parameters):

```C
#define hash_index(h)                 ((h)&hashsize1)                      /* compute chunkarr[] index for given hash value h;  we prefer to use lower bits since higher ones may be shared with stored_hash value */

#define chunkarr_value(hash,chunk)    ((Chunk(hash)&hash_mask)+(chunk))    /* combine hash and number of chunk into the one Chunk value for storing in the chunkarr[] */
#define get_hash(value)               ((value)&hash_mask)                  /* get hash from the combined value */
#define get_chunk(value)              ((value)&chunknum_mask)              /* get chunk number from the combined value */
```

And a longer sample, sorry for russian comments:
```C
  // Маска для значения, сохраняемого в хеш-таблице, позволяющая извлечь достаточно бит для адресации позиции в dict[]
  const TIndex matchmask  =  (sizeof(TIndex)<8?  0  :  roundup_to_power_of(MatchBounds.dictsize,2)) - 1;
  // Маска для значения хеш-функции / в хеш-таблице, позволяющая извлечь биты, не пересекающиеся ни с hashmask, ни с matchmask
  const HTValue tagmask   =  (sizeof(TIndex)<8?  HTValue(-1)<<32  :  ~HTValue(matchmask | hashmask));
  // При sizeof(TIndex)<8 старшее слово 64-битного match_candidates_buf->hash используется как tag, а младшее & hashmask - как индекс в hasharr[]

  TIndex last_match_end = MatchBounds.bufstart;   // points to the end of last match written, we shouldn't start a new match prior to this point

  // ОСНОВНОЙ ЦИКЛ, НАХОДЯЩИЙ ПОВТОРЯЮЩИЕСЯ СТРОКИ ВО ВХОДНЫХ ДАННЫХ
  for (MatchCandidatesProvider match_candidates(L, SEGMENT_SIZE, match_candidates_buf, MatchBounds); ; )
  {
    TCandidate::Hash  hash;
    if (match_candidates.LookaheadFutureMatchCandidate (&hash)) {
      TIndex index  =  hash & hashmask; // Индекс в хеш-таблице потенциального матча спустя INMEM_PREFETCH чанков
      prefetch (hasharr[index]);        // Prefetch the hasharr[] by INMEM_PREFETCH cycles ahead
    }

    // Проверяем совпадение в лучшем на следующие L байт блоке длины SEGMENT_SIZE
    TIndex  src;                        // Позиция начала потенциального матча
    if (! match_candidates.GetNextMatchCandidate (L, &src, &hash))  return;   // Exit function if there are no more match candidates left
    TIndex  index = hash & hashmask;    // Индекс в хеш-таблице потенциального матча, уже найденного в collect_match_candidates()
    HTValue tag   = hash & tagmask;     // Тег потенциального матча

    if (src < last_match_end) {         // Проверяем совпадение только если предыдущее найденное совпадение уже кончилось
      hasharr[index] = src | tag;       // Иначе - просто заносим в хеш-таблицу этот блок
    } else {
      HTValue prev_value = src | tag;
      TIndex  match_distance0 = 0,  match_len0 = 0,  lit_len0 = 0;  // to fill with best match info

      for (NUMBER i = 0;  i < HASHARR_BUCKET_SIZE;  ++i, ++index)
      {
        // Shift the entire hasharr[] bucket backward by one position
        HTValue next_value = hasharr[index];
        hasharr[index] = prev_value;
        prev_value = next_value;

        if unlikely ((next_value & tagmask) == tag)
        {
          TIndex dest = next_value & matchmask;       // Позиция потенциального совпадения с src
.....
```

PS: If you mean that I should use *bit ops whenever possible - yeah, I found a part in my program where it should really improve the speed, thank you! ;) But the style is never an issue for these codes, it's only speed that matters.

**Comment by Bulat-Ziganshin** [comment]
But all my codes above doesn't bit ops too much. These are examples of more intensive operations, all from the single CRC manipulation module:

```C
#define update_CRC(crc,CRCTable,c)  (CRCTable[((crc)^(c)) & 0xFF] ^ ((crc)>>8))
...
  for (NUMBER i=64; i; i/=2)
    CRCTable[i] = crc = (crc >> 1) ^ (CrcPolynomial & -(crc&1));    // crc = (crc*x) % CrcPolynomial
...
  //  for( i=0; i<32; i++ ) if( crc&(1<<i) ) out ^= mat[i];
  //  for( i=0; i<32; i++,crc>>=1 ) if( crc&1 ) out ^= mat[i];
    for( i=0; i<32; i++,crc>>=1 )  out ^= (mat[i] & -(crc&1));
...
    for( uint i=0; len; i++,len>>=1 ) if( len&1 ) crc=crc_mul(crc,crc_pad[i]);
...
    // process 16 bytes in every cycle (Slicing-by-16)
    while (length >= 16)
    {
      uint32_t a0 = *current++ ^ crc;
      uint32_t a1 = *current++;
      uint32_t a2 = *current++;
      uint32_t a3 = *current++;
      crc  = Crc32Tables[ 0][ a3>>24        ] ^
             Crc32Tables[ 1][(a3>>16) & 0xFF] ^
             Crc32Tables[ 2][(a3>> 8) & 0xFF] ^
             Crc32Tables[ 3][ a3      & 0xFF] ^
             Crc32Tables[ 4][ a2>>24        ] ^
             Crc32Tables[ 5][(a2>>16) & 0xFF] ^
             Crc32Tables[ 6][(a2>> 8) & 0xFF] ^
             Crc32Tables[ 7][ a2      & 0xFF] ^
             Crc32Tables[ 8][ a1>>24        ] ^
             Crc32Tables[ 9][(a1>>16) & 0xFF] ^
             Crc32Tables[10][(a1>> 8) & 0xFF] ^
             Crc32Tables[11][ a1      & 0xFF] ^
             Crc32Tables[12][ a0>>24        ] ^
             Crc32Tables[13][(a0>>16) & 0xFF] ^
             Crc32Tables[14][(a0>> 8) & 0xFF] ^
             Crc32Tables[15][ a0      & 0xFF];
      length -= 16;
    }
```

**Comment by Bulat-Ziganshin** [comment]
> Ridiculous, you can map clearBit to the most efficient instruction sequence.

No, it's impossible. x.clearBit(n) maps to `x & ~(1<<n)` which isn't even close as efficient as `x & mask` where mask is precomputed based on n (and stored in an object). It's a part of special low-level programming knowledge :)

**Comment by timotheecour** [comment]
> No, it's impossible. x.clearBit(n) maps to x & ~(1<<n) which isn't even close as efficient as x & mask where mask is precomputed based on n (and stored in an object)

just replying to this single point in isolation here:

* couldn't we achieve same with an overload that takes `n` as `static[int]`?
- [ ] also IIUC, the order of arguments in `proc clearBit` is flipped, and probably should be fixed to allow `x.clearBit(n)`:
`proc clearBit(nr: int; dest: var uint32) {.inline.} =...`

**Comment by Bulat-Ziganshin** [comment]
@timotheecour I talk about situation when n is computed at run-time.

OTOH, when n is fixed, good compiler should map it to AND asm operation, and Nim probably does it - one way or other (i.e. it should be hard to encode operation in such a way that C compiler can't optimize it :) )

**Comment by awr1** [comment]
I've never really had much problem with `or`/`and`/etc. working for both integers and booleans. In cases where you have
```nim
if someInlineBitwiseNonsenseGoesHere or someBool and someOtherBool:
```
a healthy dose of parentheses or moving stuff into variables should be done anyway (and is more than likely required for correct parsing).

Enhancing `bitops` seems like a good idea. I'd like to make a list of useful procs to be added, as the ones @Araq mentioned appear to be absent. So far in the thread:

```nim
type intrange = range[0 .. sizeof(int) * 8]
proc setBit(x: int; at: intrange): int
proc clearBit(x: int; at: intrange): int
proc flipBit(x: int; at: intrange): int
```
Suggesting a few others based on the use cases in this thread:
```nim
proc intersects(x, y: int): bool # x and y != 0
proc xor(n: varargs[int]): int   # n[0] xor n[1] xor n[2] .. (lisp-ish, non-operator overloading, because `xor`() is limited to two arguments)
proc or(n: varargs[int]): int    # ditto
proc and(n: varargs[int]): int   # ditto
```

**Comment by awr1** [comment]
With the last three procs, your example of a computation of a CRC32 hash @Bulat-Ziganshin could be flattened to:
```nim
 crc  = xor(Crc32Tables[ 0][ a3 shl 24          ],
            Crc32Tables[ 1][(a3 shl 16) and 0xFF],
            Crc32Tables[ 2][(a3 shl 8)  and 0xFF],
            Crc32Tables[ 3][ a3         and 0xFF],
            # ...
```
in lieu of a repeated usage of `xor`

**Comment by Bulat-Ziganshin** [comment]
@awr1 yes, as far as it generates the same code... in particular, I bet that your last definitions of and/or/xor should be templates to generate the same code as `x xor y xor z...`.

**Comment by Araq** [comment]
> No, it's impossible.

We're talking past each other a bit, let's just agree to disagree.

**Comment by andreaferretti** [comment]
One thing that I like about Nim are operations named sanely (e.g. `or`, `countLeadingZeroBits`) rather than with operators that I have to lookup each time.

@Bulat-Ziganshin precomputation can be done using consts (or even as an optimization by the c backend - gone are the days where one would need to use, say, bit shifts in place of division by powers of two)

**Comment by dom96** [comment]
Why are we even discussing this? It was made clear by @Araq that it's far too late to change this now.

**Comment by krux02** [comment]
Here my two cents, but before saying anything: This discussion can be closed. 

Introducing new operators for bit operations, I am totally against the idea. Operators are usually hard to look up and heavily overloaded. They should be used very carefully. I would even say, if there isn't a reference to use the operator in the paper world, then the operator should be rejected in the Nim standard library.  The C programming language gave some bit operations an operator but others are not available, even though they have a corresponding assembly instruction. 

I also think it is too late to change the meaning of ``or`` on integers.  
But if it would not be to late, I would probably give the the C style bit operations proper function names in the bitops module.

**Comment by mratsim** [comment]
I'm probably one of the most heavy users of Generics/Static + bitwise operations, here is my take on this:

1. No operator, it's a pain to search (arguably `or` and `and` are also very hard to search).
2. You can create your own templates:
  ```Nim
  template `|`(x, y: SomeNumber): SomeNumber = x or y
  ```
3. I'm not really fond of `bitOr`, `bitAnd`, `bitNot` because it's a bit noisy but I can live with it.
4. Though I don't see where it's super error-prone, there is a type system.
5. One of the appeal of Nim is same application as C but with modern features. This is especially important for embedded. Bit operations should stay very ergonomic.
6. Operator precedence should not be changed without a great deal of thought and a long testing period, it's too hard to debug in existing codebases. For example, a reasonable syntax to slice array with a step of 2 would be `a[0..10|2]` but this is read as `a[0..(10|2)]`. Suddenly changing precedence means implementing or deleting lots of workarounds and tests to make sure there is no surprise.
7. I'd like to avoid taking arrows like `>>=` and `>>>`. Compilers can optimize properly `x = x shr n`, however arrows are quite valuable for functional programming (`>>=` is bind in Haskell), well the priority is also wrong for many like `>>=`, `<$>` and `<*>` and probably more. I feel taking those might prevent great usage by libraries down the road.

**Comment by Bulat-Ziganshin** [comment]
@mratsim 
> I'd like to avoid taking arrows like >>= and >>>.

we are probably discussing what to put into extra package. You should just avoid importing Haskell monads and low-level bitops into the same module

> Compilers can optimize properly x = x shr n, 

`x = x shr n` may be hard to optimize when x is a complex expression. Or, better said, there are no guarantees of optimization, and I need guarantees to write high-optimized code.

> well the priority is also wrong

it's about user-defined op priorities, again

> Operator precedence should not be changed without a great deal of thought and a long testing period

since `| &` are already taken by other ops in system.nim, I think that their chances to be reused for bit ops are zero

> No operator, it's a pain to search

I don't search in my C++ code for `|` as well as I don't search for `*` or `+`. Your needs may vary.

**Comment by timotheecour** [comment]
@araq  @dom96 @mratsim
> Why are these "super error prone" and a problem for "generic programming"? I don't like them too much either but have yet to see a bug that isn't caught by Nim's type checker on these.

It's error prone when user thinks code has logical semantic but compiler interprets it as bitwise semantic, eg:
```
import osproc
echo execCmd("grep CAN_DELETE foo.txt") and execCmd("rm foo.txt") or execCmd("rm foo.backup.txt")
```

user may wrongly thinks this would map to (in bash):

`grep CAN_DELETE foo.txt && rm foo.txt || rm foo.backup.txt`

which would **delete only 1 file**; however the code compiles and runs and **deletes both files**. 

Of course in this simplified example the user code has a bug (should be `execCmd("...") == 0`), however, this bug would've been caught under my proposal with a type mismatch error instead of silently accepting it with bitwise semantics.

I can't think of real life example where you'd have a generic that operates as either logical or bitwise depending on type. 

Finally another argument: logical `and`, `or` short circuit (as in this bash example), but not bitwise arguments; it's really odd to conflate these 2 under same name.

**Comment by mratsim** [comment]
@Bulat-Ziganshin 

> `x = x shr n` may be hard to optimize when x is a complex expression. Or, better said, there are no guarantees of optimization, and I need guarantees to write high-optimized code.

I don't see how it can be a complex expression, at a low-level x must be a pointer so the compiler must have resolved that to a memory address so to do anything the compiler must transform that to `valueAtAddress = valueAtAddress << n` in its own intermediate representation.

Now I agree that it had always bugged me to not have in-place bitwise ops.

**Comment by timotheecour** [comment]
> Ok, so you mean the adhoc overloading in this case is not a benefit for generic programming. But it doesn't cause problems either. (Unless you only pass the value to an "echo".)

or `discard` or a generic (or overloaded proc) that would accept either.

**Comment by ReneSac** [comment]
I had a formely named "bithacks" but now that I removed what bitops implemented I renamed it to "c_alikes". Sorry for keeping it private all this time, as I went on and off nim and never really finished it like I wanted. I did published a few parts of it via gists circa 2015, but now I finally cleaned it up and published as a proper nimble package:  https://github.com/ReneSac/c_alikes

As the name implies, I tried to make those operators resembling C/C++ (that is certainly the language where most bitwise code is written), but with compromises to work in vanilla Nim. I discussed with Araq at the time about the operators. The `*>` and `*<` instead of `>>` and `<<` for the correct precedence and for the in-place version `*>=` to work and not be considered a comparison operator were Araq's ideas.

I feel the priority for the `&` and `|` operators in Nim is backwards, but is too late to change that, and I think it is better to live with that slightly off priorities than Araq's suggestion of equalizing both priorities via `|&` instead of just `&`. I did adopt that suggestion for bitwise `xor`: `|^` both for priority reasons and because `^` alone was already taken. Anyway, those priorities are already much better than copying the priority of the logical `and` and `or`, and is a place where C shouldn't be copied.

I also mixed in that module functions discussed here like clearBit, etc, as well as pointer arithmetic and a few other misc things.

@timotheecour: that way of limiting the operations to inside a block doesn't really work as a library. Besides being a pita to actually use, Nim will complain that procs/templates must be defined at the top level to be exported.
